### PR TITLE
Fix reconnection and retries

### PR DESCRIPTION
**Pull request by yash-dk**
This is to identify the issue #1556 and partially #1541.
The main error was is in 
`def retry_range(retries):`

### CODE REVIEW

File: telethon/helpers.py [hunk 1]
```diff
@@ -101,11 +101,11 @@ def retry_range(retries):
     not a zero or a positive integer value, the sequence will be
     infinite, otherwise it will end at `retries + 1`.
     """
-    yield 1
```
Comment (Lonami) [inline telethon/helpers.py:104]: The problem with this is that it's *retries*. For 0 retries, it should try once (1st try) and then retry no more (@cher-nov correct me if I'm wrong, but I think this was our rationale?). Now with 0 retries we won't even try to, say, connect, because the loop won't iterate even once.

File: telethon/network/mtprotosender.py [hunk 2]
```diff
@@ -357,32 +358,36 @@ async def _reconnect(self, last_error):
...
 
         retries = self._retries if self._auto_reconnect else 0
-        for attempt in retry_range(retries):
-            try:
-                await self._connect()
-            except (IOError, asyncio.TimeoutError) as e:
-                last_error = e
-                self._log.info('Failed reconnection attempt %d with %s',
-                               attempt, e.__class__.__name__)
+        if retries:
```
Comment (Lonami) [inline telethon/network/mtprotosender.py:361]: Separating the range into its own method was exactly made to avoid this kind of extra works that needs to be done outside and keep the logic inside the method itself.

**Review by Lonami — COMMENTED** [review]
I still haven't checked the problem in depth, but maybe it's because we're using `retry_range` to both try first time and start retrying? With an example:

```python
# good, we need a first attempt to connect but 0 retries means we won't retry connecting
for attempt in retry_range(retries):
    connect()

# bad, we don't want to try reconnecting but 0 will still try once
for attempt in retry_range(retries):
    reconnect()
```

File: telethon/helpers.py (hunk 1 above)
Comment (yash-dk) [reply telethon/helpers.py:104]: Yes, that's right that was the reason to yield 1 before while loop.
But this is breaking for the reconnects as you mentioned before.

File: telethon/network/mtprotosender.py (hunk 2 above)
Comment (yash-dk) [reply telethon/network/mtprotosender.py:361]: Ok, now I get it maybe we can just take an extra parameter for `retry_range` for reconnecting and init the attempt before the loop.

**Comment by yash-dk** [comment]
Can you check this implementation?

File: telethon/helpers.py (hunk 1 above)
Comment (cher-nov) [reply telethon/helpers.py:104]: > @cher-nov correct me if I'm wrong, but I think this was our rationale?
Literally this, yes.

File: telethon/network/mtprotosender.py [hunk 3]
```diff
@@ -384,6 +384,7 @@ async def _reconnect(self, last_error):
             self._log.error('Automatic reconnection failed %d time(s)', attempt)
             await self._disconnect(error=last_error.with_traceback(None))
 
+
```
Comment (cher-nov) [inline telethon/network/mtprotosender.py:387]: Any reason for whitespace change?

_[№ 3]_

File: telethon/network/mtprotosender.py [hunk 4]
```diff
@@ -357,25 +358,24 @@ async def _reconnect(self, last_error):
         self._state.reset()
 
         retries = self._retries if self._auto_reconnect else 0
-        for attempt in retry_range(retries):
+        
+        attempt = 0
+        for attempt in retry_range(retries,reconnect=True):
```
Comment (cher-nov) [inline telethon/network/mtprotosender.py:363]: Missed space before `reconnect` argument.
Also don't use explicit naming of arguments unless it's really necessary.

File: telethon/network/mtprotosender.py [hunk 5]
```diff
@@ -217,6 +217,7 @@ async def _connect(self):
         self._log.info('Connecting to %s...', self._connection)
 
         connected = False
+        attempt = 0
```
Comment (Lonami) [inline telethon/network/mtprotosender.py:220]: This assignment serves no purpose.

File: telethon/network/mtprotosender.py (hunk 4 above)
Comment (Lonami) [inline telethon/network/mtprotosender.py:363]:
```suggestion
        for attempt in retry_range(retries, reconnect=True):
```

**Review by Lonami — CHANGES_REQUESTED** [review]

File: telethon/helpers.py [hunk 6]
```diff
@@ -95,17 +95,20 @@ def strip_text(text, entities):
...
+
+    if retries == 0 and not reconnect:
+        yield 1
+
     attempt = 0
     while attempt != retries:
-        attempt += 1
         yield 1 + attempt
+        attempt += 1
```
Comment (cher-nov) [inline telethon/helpers.py:111]: This implementation is wrong: it will yield `1` twice in case of `retry_range(0, False)`.

File: telethon/client/users.py [hunk 7]
```diff
@@ -51,6 +51,7 @@ async def _call(self: 'TelegramClient', sender, request, ordered=False):
 
         request_index = 0
         self._last_request = time.time()
+        attempt = 0
         for attempt in retry_range(self._request_retries):
```
Comment (cher-nov) [inline telethon/client/users.py:55]: Seems pretty like an anti-pattern for me, though I see why you've added `attempt` assignment there.
Can it be rewritten without it? If no, then at least add a whitespace line before assignment.

**Review by cher-nov — CHANGES_REQUESTED** [review]
Thank you for the effort, but I'm pretty unsure at the moment that it's a correct fix.

File: telethon/helpers.py (hunk 6 above)
Comment (yash-dk) [reply telethon/helpers.py:111]: No, it won't, as `attempt != retries` will be false at the first stage and thus never yield an extra 1.

File: telethon/helpers.py [hunk 8]
```diff
@@ -95,17 +95,20 @@ def strip_text(text, entities):
     return text
 
 
-def retry_range(retries):
+def retry_range(retries, reconnect=False):
```
[the remark this answers is not in the document]
Comment (yash-dk) [reply telethon/helpers.py:98]: maybe force_retry ?

File: telethon/network/mtprotosender.py [hunk 9]
```diff
@@ -357,25 +358,24 @@ async def _reconnect(self, last_error):
...
                 self._log.info('Failed reconnection attempt %d with %s',
-                               attempt, e.__class__.__name__)
-
+                            attempt, e.__class__.__name__)
                 await asyncio.sleep(self._delay)
             except Exception as e:
                 last_error = e
                 self._log.exception('Unexpected exception reconnecting on '
                                     'attempt %d', attempt)
-
```
[the remark this answers is not in the document]
Comment (yash-dk) [reply telethon/network/mtprotosender.py:373]: fixed, got added while testing stuff. Same for all these kinds of issues.

File: telethon/helpers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yash-dk) [reply telethon/helpers.py:98]: This way it will try to re_connect 1 time even if the retries are set to 0.
But using the special parameter will ensure that the reconnect is not initiated. 
We can assign None if retries are 0 above the loop but this will contradict the below I guess?
The logic will be inside the helper func as lonami said.

File: telethon/helpers.py (hunk 6 above)
Comment (cher-nov) [reply telethon/helpers.py]: Ah, my bad.

File: telethon/helpers.py [hunk 10]
```diff
@@ -95,16 +95,19 @@ def strip_text(text, entities):
...
     infinite, otherwise it will end at `retries + 1`.
     """
 
-    if retries == 0 and not reconnect:
+    # We need atleast one iteration even if the retries are 0
+    # when force_retry is True. 
+    if retries == 0 and force_retry:
         yield 1
 
+    # If retries are non 0 then itetrate
```
Comment (cher-nov) [inline telethon/helpers.py:110]: Typo: `itetrate`, must be `iterate`.

File: telethon/helpers.py [hunk 11]
```diff
@@ -95,16 +95,19 @@ def strip_text(text, entities):
...
     attempt = 0
     while attempt != retries:
         yield 1 + attempt
```
Comment (cher-nov) [inline telethon/helpers.py:113]: You didn't fix this as I said, instead, you've made a redundant increment.

**Review by cher-nov — CHANGES_REQUESTED** [review]

File: telethon/helpers.py (hunk 11 above)
Comment (yash-dk) [reply telethon/helpers.py]: I changed it to -

```
attempt += 1
yield attempt
```

and the other Special one I stated the reason

File: telethon/helpers.py [hunk 12]
```diff
@@ -95,17 +95,24 @@ def strip_text(text, entities):
...
-def retry_range(retries):
+def retry_range(retries, force_retry=True):
     """
     Generates an integer sequence starting from 1. If `retries` is
     not a zero or a positive integer value, the sequence will be
     infinite, otherwise it will end at `retries + 1`.
     """
-    yield 1
+
+    # We need atleast one iteration even if the retries are 0
```
Comment (Lonami) [inline telethon/helpers.py:105]:
```suggestion
    # We need at least one iteration even if the retries are 0
```

File: telethon/helpers.py [hunk 13]
```diff
@@ -95,17 +95,24 @@ def strip_text(text, entities):
...
+    # when force_retry is True. 
+    if retries == 0 and force_retry:
+        yield 1
+
+    # If retries are non 0 then iterate
     attempt = 0
     while attempt != retries:
         attempt += 1
-        yield 1 + attempt
+        yield attempt
```
Comment (Lonami) [inline telethon/helpers.py:114]: This will yield 1 twice. Perhaps moving `attempt` a bit earlier would make this cleaner:

```python
attempt = 0
if retries == 0 and force_retry:
    attempt += 1
    yield attempt

# If retries are non 0 then iterate
while ...:
```

File: telethon/network/mtprotosender.py [hunk 14]
```diff
@@ -357,14 +358,15 @@ async def _reconnect(self, last_error):
         self._state.reset()
 
         retries = self._retries if self._auto_reconnect else 0
-        for attempt in retry_range(retries):
+        
+        attempt = 0
+        for attempt in retry_range(retries, False):
```
Comment (Lonami) [inline telethon/network/mtprotosender.py:363]:
```suggestion
        # We're already "retrying" to connect, so we don't want to force retries
        for attempt in retry_range(retries, force_retry=False):
```

File: telethon/network/mtprotosender.py [hunk 15]
```diff
@@ -357,14 +358,15 @@ async def _reconnect(self, last_error):
...
             try:
                 await self._connect()
             except (IOError, asyncio.TimeoutError) as e:
                 last_error = e
                 self._log.info('Failed reconnection attempt %d with %s',
-                               attempt, e.__class__.__name__)
-
+                                attempt, e.__class__.__name__)
```
Comment (Lonami) [inline telethon/network/mtprotosender.py:369]: Could you get revert this change? Previously it was properly aligned with `(`, but it is no longer aligned (it's also not relevant to this PR).

**Review by Lonami — CHANGES_REQUESTED** [review]

File: telethon/helpers.py (hunk 13 above)
Comment (yash-dk) [reply telethon/helpers.py:114]: it will not yield 1, 2 times
as `attempt != retries` will be false at the first stage and thus never yield an extra 1.

File: telethon/network/mtprotosender.py (hunk 14 above)
Comment (yash-dk) [reply telethon/network/mtprotosender.py:363]: I used named args but was told to depricate it.

File: telethon/helpers.py (hunk 13 above)
Comment (Lonami) [reply telethon/helpers.py:114]: For this snippet:

```python
print('0 retries, force', list(retry_range(0)))
print('1 retries, force', list(retry_range(1)))
print('2 retries, force', list(retry_range(2)))
```

The old code prints:

```
print('0 retries, force', list(retry_range(0)))
print('1 retries, force', list(retry_range(1)))
print('2 retries, force', list(retry_range(2)))
```

However, with the current changes, the code now prints (including `force=False`):

```
0 retries, force [1]
1 retries, force [1]
2 retries, force [1, 2]
0 retries, no force []
1 retries, no force [1]
2 retries, no force [1, 2]
```

Note that this behavior is not correct when forcing to retry, since it yields 1 twice.

Comment (cher-nov) [reply telethon/helpers.py:114]: Ha, I was right! %-)

File: telethon/helpers.py [hunk 16]
```diff
@@ -95,17 +95,23 @@ def strip_text(text, entities):
...
     Generates an integer sequence starting from 1. If `retries` is
     not a zero or a positive integer value, the sequence will be
     infinite, otherwise it will end at `retries + 1`.
     """
-    yield 1
+
+    # We need at least one iteration even if the retries are 0
+    # when force_retry is True.
+    if force_retry:
+        retries+=1
```
Comment (Lonami) [inline telethon/helpers.py:108]: Minor style nit:

```suggestion
        retries += 1
```

**Review by Lonami — CHANGES_REQUESTED** [review]

File: telethon/helpers.py [hunk 17]
```diff
@@ -95,17 +95,23 @@ def strip_text(text, entities):
...
     """
     Generates an integer sequence starting from 1. If `retries` is
     not a zero or a positive integer value, the sequence will be
     infinite, otherwise it will end at `retries + 1`.
     """
-    yield 1
+
+    # We need at least one iteration even if the retries are 0
+    # when force_retry is True.
+    if force_retry:
```
Comment (Lonami) [inline telethon/helpers.py:107]: I just realized [the documentation states](https://docs.telethon.dev/en/latest/modules/client.html#telethon.client.telegrambaseclient.TelegramBaseClient):

> May be set to a negative or None value for infinite retries

Now using -1 will become 0 and perform 0 iterations instead of infinite. Worse, `None` will crash. We need to modify this to account for that.

```suggestion
    if force_retry and not (retries is None or retries < 0):
```

**Review by Lonami — CHANGES_REQUESTED** [review]

File: telethon/helpers.py (hunk 17 above)
Comment (yash-dk) [reply telethon/helpers.py]: Got it.
